### PR TITLE
HOTFIX: Fix put logic

### PR DESCRIPTION
**Pull request by enothereska**

### CODE REVIEW

**Comment by enothereska** [comment]
@dguy @guozhangwang please have a look.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/ThreadCache.java [hunk 1]
```diff
@@ -122,13 +122,17 @@ public void put(final String namespace, byte[] key, LRUCacheEntry value) {
     }
 
     public LRUCacheEntry putIfAbsent(final String namespace, byte[] key, LRUCacheEntry value) {
-        final NamedCache cache = getOrCreateCache(namespace);
-        return cache.putIfAbsent(Bytes.wrap(key), value);
+        final LRUCacheEntry originalValue = get(namespace, key);
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/ThreadCache.java]: This method is not `synchronized`. So there could be a race condition here. I think it should be:

```
final NamedCache cache = getOrCreateCache(namespace);
final LRUCacheEntry result = cache.putIfAbsent(Bytes.wrap(key), value);
maybeEvict(namespace);
return result;
```

**Review by dguy — CHANGES_REQUESTED** [review]
Maybe add a couple of tests to `ThreadCacheTest` to prove that items get evicted when `putAll` or `putIfAbsent` are called?

**Comment by enothereska** [comment]
Thanks @dguy

**Review by dguy — APPROVED** [review]
LGTM

**Comment by guozhangwang** [comment]
LGTM.

**Comment by guozhangwang** [comment]
Merged to trunk.

**Comment by guozhangwang** [comment]
@enothereska Could you file a follow-up PR for 0.10.1 branch just in case we have a bug-fix release later?

**Pull request closed without merging**